library/math/dual.lisp

```diff
@@ -0,0 +1,177 @@
+    (Dual :t :t))
+   
+  (declare primal-part (Dual :t -> :t))
+  (define (primal-part (Dual p _))
+    p)
+
+  (declare dual-part (Dual :t -> :t))
+  (define (dual-part (Dual _ d))
+    d)
+
+  (define (sq x)
+    (* x x))
+   
+  (define-instance (Eq :t => Eq (Dual :t))
+    "Note: Eq only works on the primal components."
+    (define (== (Dual a _) (Dual p _))
+      (== a p)))
+  
+  (define-instance (Num :t => Num (Dual :t)) 
+    (define (+ (Dual p1 d1) (Dual p2 d2))
+      (Dual (+ p1 p2) (+ d1 d2)))
+     
+    (define (- (Dual p1 d1) (Dual p2 d2))
+      (Dual (- p1 p2) (- d1 d2)))
+
+    (define (* (Dual p1 d1) (Dual p2 d2))
+      (Dual (* p1 p2)
+            (+ (* p1  d2) (* d1 p2))))
+     
+    (define (fromInt z)
+      (Dual (fromInt z) 0)))
+   
+  (define-instance (Reciprocable :t => Reciprocable (Dual :t))
+    (define (/ (Dual p1 d1) (Dual p2 d2))
+      (Dual (/ p1 p2)
+            (/ (- (* d1 p2)
+                  (* p1 d2))
+  	       (* p2 p2))))
+     
+    (define (reciprocal (Dual p1 d1))
+      (Dual (reciprocal p1)
+            (/ (negate d1) (* p1 p1)))))
+
+  (define-instance ((Num :t) (Trigonometric :t) (Reciprocable :t) (Radical :t) => (Trigonometric (Dual :t)))
+    (define (sin (Dual p1 d1))
+      (Dual (sin p1)
+            (* d1 (cos p1))))
+     
+    (define (cos (Dual p1 d1))
+      (Dual (cos p1)
+            (negate (* d1 (sin p1)))))
+     
+    (define (tan (Dual p1 d1))
+      (Dual (tan p1)
+            (/ d1 (sq (cos p1)))))
+     
+    (define (asin (Dual p1 d1))
+      (Dual (asin p1)
+            (/ d1 (sqrt (- 1 (sq p1))))))
+     
+    (define (acos (Dual p1 d1))
+      (Dual (acos p1)
+            (negate (/ d1 (sqrt (- 1 (sq p1)))))))
+     
+    (define (atan (Dual p1 d1))
+      (Dual (atan p1)
+            (/ d1 (+ 1 (sq p1))))))
+
+  (define-instance ((Num :t) (Exponentiable :t) (Reciprocable :t) => (Exponentiable (Dual :t)))
+    (define (exp (Dual p1 d1))
+      (Dual (exp p1)
+            (* d1 (exp p1))))
+
+    (define (pow dual1 dual2)
+      (exp (* dual2 (ln dual1))))
+     
+    (define (ln (Dual p1 d1))
+      (Dual (ln p1)
+            (/ d1 p1)))
+     
+    (define (log dual1 dual2)
+      (/ (ln dual2) (ln dual1))))
+   
+  (define-instance ((Num :t) (Radical :t) (Reciprocable :t) (Exponentiable :t) => (Radical (Dual :t)))
+    (define (nth-root n (Dual p1 d1))
+      (let ((n* (fromInt n)))
+        (Dual (nth-root n p1)
+  	      (pow (* p1 (/ d1 n*)) (- (reciprocal n*) 1)))))
+     
+    (define (sqrt (Dual p1 d1))
+      (Dual (sqrt p1)
+            (/ p1 (* 2 (sqrt p1))))))
+
+  (define-instance ((Ord :t) (Ord :t) => Ord (Dual :t))
+    "Note: Ord only works on the primal components."
+    (define (<=> (Dual p1 _) (Dual p2 _))
+      (<=> p1 p2)))
+      
+  (define-instance ((Hash :t) (Hash :t) => (Hash (Dual :t)))
+    "Note: Hash only works on the primal component."
+    (define (hash (Dual p1 _))
+      (hash p1))))
```

Comment:
I would actually recommend mixing a constant hash value in with this here, so that the hash value of `(Dual :t)` is explicitly different from `:t`. Look at how other hash functions are implemented to see how.

Reply:
I don't think we should mix in a constant. We have lots of other cross type hash collisions.

Reply:
Not doing it saves two or three assembly instructions.

Doing it makes debugging ever so marginally easier.

Reply:
It makes hash-tables asymptotically worse by reducing entropy in the hash. We [have discusses this previously](https://github.com/coalton-lang/coalton/pull/357#issuecomment-1006241282).

Reply:
If we do hash type information here, we should probably do it everywhere else for consistency.

Reply:
Given that we don't (intend to) support mixed-type hash tables, I see no reason to mix type ids into hashes. In dynamically typed contexts, it's valuable to make sure that e.g. the double-float `0d0` hashes differently from the single-float `0f0`, but in Coalton that doesn't matter; those values won't co-exist as keys of the same hash table. And @eliaslfox is right; every additional value you mix into a hash reduces the "weight" of each other value in the resulting hash, so mixing in a constant reduces overall entropy. I doubt that the entropy loss matters significantly for numeric types except for very large hash-tables, but there's no reason at all to mix in a constant, as compared to a small reason not to.

Reply:
1. Coalton still lives inside of Common Lisp and is debugged inside of Common Lisp. I hope we don't forget that.

2. I don't understand why you believe that mixing a constant reduces entropy. Even if the hash were `hash(x)+C mod 2^n`, we still have the same information content, for we can always unmix this.

Reply:
Are we prepared to assume that `mix-hashes` is invertible? I know that `logxor` is, which I think we use as a fallback, and I haven't checked whether SBCL's is, but I know that for many hashing algorithms it is not. For example, I recently implemented FxHash, which uses a `wrapping-mul` and is therefore not invertible.

I could be doing my math wrong, but it seems to me that hashing a value of `N` bits into a hash of `H` bits captures `(/ H N)` bits of entropy from each bit of the value. Adding a constant of `C` bits means you instead capture `(/ H (+ N C))` bits from each bit of the value, which is less than `(/ H N)`.

Reply:
I don't know about `mix-hashes` but I'm not advocating for any function in particular, just one that lets us make hash codes for different object types distinguished.

It's all very minor which is why I let go of it before, and I'm fine to let go of it now.